amr-wind/utilities/diagnostics.cpp

```diff
@@ -1,8 +1,196 @@
+                for (int n = 0; n < ncomp; ++n) {
+                    min_fab = amrex::min(min_fab, f_arr(i, j, k, n));
+                }
+            });
+            return min_fab;
+        });
+}
+
+void amr_wind::diagnostics::get_mfab_extrema(
+    amrex::Real& mfab_max_val,
+    amrex::Real& mfab_min_val,
+    const amrex::MultiFab& mfab,
+    const amrex::MultiFab& mfab_mask,
+    const amrex::Real mask_val,
+    const int nghost)
+{
+    int ncomp = mfab.n_comp;
+    mfab_max_val = amrex::ReduceMax(
+        mfab, mfab_mask, nghost,
+        [=] AMREX_GPU_HOST_DEVICE(
+            amrex::Box const& bx, amrex::Array4<amrex::Real const> const& f_arr,
+            amrex::Array4<amrex::Real const> const& mask_arr) -> amrex::Real {
+            amrex::Real max_fab = -1e8;
+            amrex::Loop(bx, [=, &max_fab](int i, int j, int k) noexcept {
+                for (int n = 0; n < ncomp; ++n) {
+                    max_fab = amrex::max(
+                        max_fab, std::abs(mask_arr(i, j, k) - mask_val) <
+                                         constants::TIGHT_TOL
+                                     ? f_arr(i, j, k, n)
+                                     : max_fab);
+                }
+            });
+            return max_fab;
+        });
+
+    mfab_min_val = amrex::ReduceMin(
+        mfab, mfab_mask, nghost,
+        [=] AMREX_GPU_HOST_DEVICE(
+            amrex::Box const& bx, amrex::Array4<amrex::Real const> const& f_arr,
+            amrex::Array4<amrex::Real const> const& mask_arr) -> amrex::Real {
+            amrex::Real min_fab = 1e8;
+            amrex::Loop(bx, [=, &min_fab](int i, int j, int k) noexcept {
+                for (int n = 0; n < ncomp; ++n) {
+                    min_fab = amrex::min(
+                        min_fab, std::abs(mask_arr(i, j, k) - mask_val) <
+                                         constants::TIGHT_TOL
+                                     ? f_arr(i, j, k, n)
+                                     : min_fab);
+                }
+            });
+            return min_fab;
+        });
+}
+
+void amr_wind::diagnostics::make_mask_multiplier(
+    amrex::MultiFab& mfab,
+    const amrex::MultiFab& mfab_mask,
+    const amrex::Real mask_val,
+    const amrex::Real set_val)
+{
+    const auto& arr = mfab.arrays();
+    const auto& arr_mask = mfab_mask.const_arrays();
+    amrex::ParallelFor(
+        mfab, mfab.n_grow, mfab.n_comp,
+        [=] AMREX_GPU_DEVICE(int nbx, int i, int j, int k, int n) noexcept {
+            arr[nbx](i, j, k, n) = std::abs(arr_mask[nbx](i, j, k) - mask_val) <
+                                           constants::TIGHT_TOL
+                                       ? 1.0
+                                       : set_val;
+        });
+}
+
+void amr_wind::diagnostics::get_field_extrema(
+    amrex::Real& field_max_val,
+    amrex::Real& field_min_val,
+    const amr_wind::Field& field,
+    const int icomp,
```

Comment:
Only thing I would change, just to be consistent with other amrex-like things is to specify the start and end component instead of having logic on `icomp`.

Reply:
good idea. I think I'll do the starting component and the number of components, like in the MultiFab Multiply function